File: COCI '23 Contest 1 #3 AN2DL.py
# TLE, PYTHON IS TOO SLOW, CHECK C++ CODE
# https://dmoj.ca/problem/coci23c1p3
# create N monotonic queues for each row
# we can shift the columns in O(1)
# we can also get the max from the rows in O(n)
#
# mostly implementation

from collections import deque
import sys

input = sys.stdin.readline

N, M = map(int, input().split())
grid = [list(map(int, input().split())) for _ in range(N)]
r, c = map(int, input().split())


def fill_start(row):
    """beginning state of queues for each column"""
    q = deque()
    for i, v in enumerate(row[:c]):
        while q and row[q[-1]] <= v:
            q.pop()
        q.append(i)
    return q


def solve_col(max_i):
    """solves an entire column, given the max in each grid[r][col] for all r"""
    max_vals = [grid[rr][i] for rr, i in enumerate(max_i)]
    q = deque()
    res = [0] * (N - r + 1)
    for i, v in enumerate(max_vals):
        left = i - r + 1
        while q and max_vals[q[-1]] <= v:
            q.pop()
        q.append(i)

        if left >= 0:  # full window, start filling values
            res[left] = max_vals[q[0]]

        if q[0] == left:  # left side exits window
            q.popleft()

    return res


# fill up starting state of column queues
maxs = [fill_start(grid[i]) for i in range(N)]  # deques are decreasing as we want to get the largest element

total_res = []
total_res.append(solve_col(list(map(lambda q: q[0], maxs))))

for j in range(c, M):
    for i in range(N):  # shift each col queue forward by 1
        row = grid[i]
        q = maxs[i]
        if q[0] == j - c:  # left element exits
            q.popleft()

        while q and grid[i][j] >= row[q[-1]]:
            q.pop()
        q.append(j)

    total_res.append(solve_col(list(map(lambda q: q[0], maxs))))  # get the rows max based on columns max

for i in list(zip(*total_res)):  # row <-> col
    print(" ".join(map(str, i)))

File: COCI '23 Contest 1 #3 AN2DL.cpp
// C++ translation that passes

#include <bits/stdc++.h>
using namespace std;

int main() {
    ios_base::sync_with_stdio(false);
    cin.tie(NULL);

    int N, M;
    cin >> N >> M;
    vector<vector<int>> grid(N, vector<int>(M));
    for(int i = 0; i < N; i++)
        for(int j = 0; j < M; j++)
            cin >> grid[i][j];

    int r, c;
    cin >> r >> c;

    vector<deque<int>> maxs(N);
    for(int i = 0; i < N; i++) {
        for(int j = 0; j < c; j++) {
            while(!maxs[i].empty() && grid[i][maxs[i].back()] <= grid[i][j])
                maxs[i].pop_back();
            maxs[i].push_back(j);
        }
    }

    vector<vector<int>> total_res;
    for(int j = c-1; j < M; j++) {
        if(j >= c) {
            for(int i = 0; i < N; i++) {
                if(maxs[i].front() == j - c)
                    maxs[i].pop_front();
                while(!maxs[i].empty() && grid[i][maxs[i].back()] <= grid[i][j])
                    maxs[i].pop_back();
                maxs[i].push_back(j);
            }
        }

        vector<int> max_vals(N);
        for(int i = 0; i < N; i++)
            max_vals[i] = grid[i][maxs[i].front()];

        deque<int> q;
        vector<int> res(N - r + 1);
        for(int i = 0; i < N; i++) {
            int left = i - r + 1;
            while(!q.empty() && max_vals[q.back()] <= max_vals[i])
                q.pop_back();
            q.push_back(i);

            if(left >= 0) {
                res[left] = max_vals[q.front()];
                if(q.front() == left)
                    q.pop_front();
            }
        }

        total_res.push_back(res);
    }

    for(int i = 0; i < total_res[0].size(); i++) {
        for(int j = 0; j < total_res.size(); j++)
            cout << total_res[j][i] << " ";
        cout << "\n";
    }

    return 0;
}
